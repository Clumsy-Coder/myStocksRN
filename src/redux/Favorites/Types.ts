import { Action } from 'redux';

/*
  ██████╗ ███████╗██████╗ ██╗   ██╗ ██████╗███████╗██████╗     ███████╗████████╗ █████╗ ████████╗███████╗
  ██╔══██╗██╔════╝██╔══██╗██║   ██║██╔════╝██╔════╝██╔══██╗    ██╔════╝╚══██╔══╝██╔══██╗╚══██╔══╝██╔════╝
  ██████╔╝█████╗  ██║  ██║██║   ██║██║     █████╗  ██████╔╝    ███████╗   ██║   ███████║   ██║   █████╗
  ██╔══██╗██╔══╝  ██║  ██║██║   ██║██║     ██╔══╝  ██╔══██╗    ╚════██║   ██║   ██╔══██║   ██║   ██╔══╝
  ██║  ██║███████╗██████╔╝╚██████╔╝╚██████╗███████╗██║  ██║    ███████║   ██║   ██║  ██║   ██║   ███████╗
  ╚═╝  ╚═╝╚══════╝╚═════╝  ╚═════╝  ╚═════╝╚══════╝╚═╝  ╚═╝    ╚══════╝   ╚═╝   ╚═╝  ╚═╝   ╚═╝   ╚══════╝

*/

/**
 * Favorites reducer state
 */
export interface FavoritesReducerState {
  symbols: string[];
}

// ---------------------------------------------------------------------------------------------- //
/*
   █████╗  ██████╗████████╗██╗ ██████╗ ███╗   ██╗    ████████╗██╗   ██╗██████╗ ███████╗███████╗
  ██╔══██╗██╔════╝╚══██╔══╝██║██╔═══██╗████╗  ██║    ╚══██╔══╝╚██╗ ██╔╝██╔══██╗██╔════╝██╔════╝
  ███████║██║        ██║   ██║██║   ██║██╔██╗ ██║       ██║    ╚████╔╝ ██████╔╝█████╗  ███████╗
  ██╔══██║██║        ██║   ██║██║   ██║██║╚██╗██║       ██║     ╚██╔╝  ██╔═══╝ ██╔══╝  ╚════██║
  ██║  ██║╚██████╗   ██║   ██║╚██████╔╝██║ ╚████║       ██║      ██║   ██║     ███████╗███████║
  ╚═╝  ╚═╝ ╚═════╝   ╚═╝   ╚═╝ ╚═════╝ ╚═╝  ╚═══╝       ╚═╝      ╚═╝   ╚═╝     ╚══════╝╚══════╝

*/

/**
 * Favorites action types
 */
export enum ActionTypes {
  ADD_FAVORITE_STOCK = 'FAVORITES/ADD_FAVORITE_STOCK',
  REMOVE_FAVORITE_STOCK = 'FAVORITES/REMOVE_FAVORITE_STOCK',
  CLEAR_FAVORITE_STOCKS = 'FAVORITES/CLEAR_FAVORITE_STOCKS',
}

// ---------------------------------------------------------------------------------------------- //
/*
   █████╗  ██████╗████████╗██╗ ██████╗ ███╗   ██╗███████╗
  ██╔══██╗██╔════╝╚══██╔══╝██║██╔═══██╗████╗  ██║██╔════╝
  ███████║██║        ██║   ██║██║   ██║██╔██╗ ██║███████╗
  ██╔══██║██║        ██║   ██║██║   ██║██║╚██╗██║╚════██║
  ██║  ██║╚██████╗   ██║   ██║╚██████╔╝██║ ╚████║███████║
  ╚═╝  ╚═╝ ╚═════╝   ╚═╝   ╚═╝ ╚═════╝ ╚═╝  ╚═══╝╚══════╝

*/

/**
 * Add stock to Favorites action
 */
export interface AddFavoriteStockAction extends Action<ActionTypes.ADD_FAVORITE_STOCK> {
  type: ActionTypes.ADD_FAVORITE_STOCK;
  readonly stockSymbol: string;
}

/**
 * Remove stock from Favorites action
 */
export interface RemoveFavoriteStockAction extends Action<ActionTypes.REMOVE_FAVORITE_STOCK> {
  type: ActionTypes.REMOVE_FAVORITE_STOCK;
  readonly stockSymbol: string;
}

/**
 * Clear all stocks from Favorites action
 */
export interface ClearFavoriteStocksAction extends Action<ActionTypes.CLEAR_FAVORITE_STOCKS> {
  type: ActionTypes.CLEAR_FAVORITE_STOCKS;
}

/**
 * Union of actions for AddFavoriteStock, RemoveFavoriteStock, ClearFavoriteStocks
 */
export type FavoritesActions =
  | AddFavoriteStockAction
  | RemoveFavoriteStockAction
  | ClearFavoriteStocksAction;
