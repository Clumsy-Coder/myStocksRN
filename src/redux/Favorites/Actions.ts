import { ActionCreator } from 'redux';

import {
  ActionTypes,
  AddFavoriteStockAction,
  RemoveFavoriteStockAction,
  ClearFavoriteStocksAction,
} from 'src/redux/Favorites/Types';

/**
 * Add stock to Favorites action creator.
 * @param stockSymbol - Stock symbol in uppercase. Ex: AAPL
 */
export const addFavoriteStock: ActionCreator<AddFavoriteStockAction> = (
  stockSymbol: string,
): AddFavoriteStockAction => ({
  type: ActionTypes.ADD_FAVORITE_STOCK,
  stockSymbol,
});

/**
 * Remove stock from Favorites action creator.
 * @param stockSymbol - Stock symbol in uppercase. Ex: AAPL
 */
export const removeFavoriteStock: ActionCreator<RemoveFavoriteStockAction> = (
  stockSymbol: string,
): RemoveFavoriteStockAction => ({
  type: ActionTypes.REMOVE_FAVORITE_STOCK,
  stockSymbol,
});

/**
 * Clear all stocks from Favorites action creator.
 */
// eslint-disable-next-line max-len
export const clearFavoriteStocks: ActionCreator<ClearFavoriteStocksAction> = (): ClearFavoriteStocksAction => ({
  type: ActionTypes.CLEAR_FAVORITE_STOCKS,
});
