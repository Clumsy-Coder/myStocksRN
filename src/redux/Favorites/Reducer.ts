import { Reducer } from 'redux';

import {
  ActionTypes,
  Reducer as FavoritesReducer,
  FavoritesActions,
} from 'src/redux/Favorites/Types';

const initialState: FavoritesReducer.FavoritesReducerState = {
  symbols: [],
};

const reducer: Reducer<FavoritesReducer.FavoritesReducerState, FavoritesActions> = (
  state = initialState,
  action,
): FavoritesReducer.FavoritesReducerState => {
  switch (action.type) {
    case ActionTypes.ADD_FAVORITE_STOCK: {
      // create a map. 'key' is the stock symbol and the 'value' is the entire FavoriteStockData object
      const map = new Map(state.symbols.map((obj) => [obj['1. symbol'], obj]));
      map.set(action.stockMetadata['1. symbol'], action.stockMetadata);
      return {
        symbols: [...map.values()],
      };
    }
    case ActionTypes.REMOVE_FAVORITE_STOCK: {
      // create a map. 'key' is the stock symbol and the 'value' is the entire FavoriteStockData object
      const map = new Map(state.symbols.map((obj) => [obj['1. symbol'], obj]));
      map.delete(action.stockSymbol);
      return {
        symbols: [...map.values()],
      };
    }
    case ActionTypes.CLEAR_FAVORITE_STOCKS: {
      return {
        symbols: [],
      };
    }
    default: {
      return state;
    }
  }
};

export default reducer;
