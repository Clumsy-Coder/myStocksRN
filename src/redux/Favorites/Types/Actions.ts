import { Action } from 'redux';

import ActionTypes from '@redux/Favorites/Types/ActionTypes';

/**
 * Add stock to Favorites action
 */
export interface AddFavoriteStockAction extends Action<ActionTypes.ADD_FAVORITE_STOCK> {
  type: ActionTypes.ADD_FAVORITE_STOCK;
  readonly stockSymbol: string;
}

/**
 * Remove stock from Favorites action
 */
export interface RemoveFavoriteStockAction extends Action<ActionTypes.REMOVE_FAVORITE_STOCK> {
  type: ActionTypes.REMOVE_FAVORITE_STOCK;
  readonly stockSymbol: string;
}

/**
 * Clear all stocks from Favorites action
 */
export interface ClearFavoriteStocksAction extends Action<ActionTypes.CLEAR_FAVORITE_STOCKS> {
  type: ActionTypes.CLEAR_FAVORITE_STOCKS;
}

/**
 * Union of actions for AddFavoriteStock, RemoveFavoriteStock, ClearFavoriteStocks
 */
export type FavoritesActions =
  | AddFavoriteStockAction
  | RemoveFavoriteStockAction
  | ClearFavoriteStocksAction;
