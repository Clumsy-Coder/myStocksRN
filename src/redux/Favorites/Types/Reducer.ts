/**
 * Favorites reducer state
 */
export interface ReducerState {
  symbols: string[];
}
