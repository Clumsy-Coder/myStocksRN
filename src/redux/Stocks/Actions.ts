/* eslint-disable max-len */
import { ActionCreator } from 'redux';

import { ActionTypes, DataDomain, Actions } from 'src/redux/Stocks/Types';

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                                  STOCK QUOTE ACTIONS CREATORS                                  //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

/**
 * Fetch stock quote action creator.
 * ONLY to be called by the front end.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 */
export const fetchStockQuote: ActionCreator<Actions.Quote.FetchAction> = (
  stockSymbol: string,
): Actions.Quote.FetchAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE,
  stockSymbol,
});

/**
 * Fetch stock quote pending action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 */
export const fetchStockQuotePending: ActionCreator<Actions.Quote.FetchPendingAction> = (
  stockSymbol: string,
): Actions.Quote.FetchPendingAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE_PENDING,
  stockSymbol,
});

/**
 * Fetch stock quote fulfilled action creator.
 * ONLY to be called by redux saga.
 * Only used internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param payload - data fetched
 */
export const fetchStockQuoteFulfilled: ActionCreator<Actions.Quote.FetchFulfilledAction> = (
  stockSymbol: string,
  payload: DataDomain.StockQuote,
): Actions.Quote.FetchFulfilledAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE_FULFILLED,
  stockSymbol,
  payload,
});

/**
 * Fetch stock quote rejected action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param error - error message
 */
export const fetchStockQuoteRejected: ActionCreator<Actions.Quote.FetchRejectedAction> = (
  stockSymbol: string,
  error: Error,
): Actions.Quote.FetchRejectedAction => ({
  type: ActionTypes.FETCH_STOCK_QUOTE_REJECTED,
  stockSymbol,
  error,
});

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                             STOCK DAILY ADJUSTED ACTIONS CREATORS                              //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

/**
 * Fetch stock Daily Adjusted action creator.
 * ONLY to be called by the front end.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param outputsize - 'compact' or 'full'
 */
export const fetchStockDailyAdj: ActionCreator<Actions.DailyAdjusted.FetchAction> = (
  stockSymbol: string,
  outputsize: 'compact' | 'full',
): Actions.DailyAdjusted.FetchAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED,
  stockSymbol,
  outputsize,
});

/**
 * Fetch stock Daily Adjusted pending action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 */
export const fetchStockDailyAdjPending: ActionCreator<Actions.DailyAdjusted.FetchPendingAction> = (
  stockSymbol: string,
): Actions.DailyAdjusted.FetchPendingAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_PENDING,
  stockSymbol,
});

/**
 * Fetch Stock Daily Adjusted fulfilled action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param payload - data fetched
 */
export const fetchStockDailyAdjFulfilled: ActionCreator<Actions.DailyAdjusted.FetchFulfilledAction> = (
  stockSymbol: string,
  payload: DataDomain.StockDailyAdj,
): Actions.DailyAdjusted.FetchFulfilledAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_FULFILLED,
  stockSymbol,
  payload,
});

/**
 * Fetch Stock Daily Adjusted rejected action creator.
 * ONLY to be called by redux saga.
 * Only used for internal use.
 * @param stockSymbol - Company stock symbol in uppercase. Ex: AAPL
 * @param error - error message
 */
export const fetchStockDailyAdjRejected: ActionCreator<Actions.DailyAdjusted.FetchRejectedAction> = (
  stockSymbol: string,
  error: Error,
): Actions.DailyAdjusted.FetchRejectedAction => ({
  type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_REJECTED,
  stockSymbol,
  error,
});

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                                 STOCK SEARCH ACTIONS CREATORS                                  //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

export const fetchStockSearch: ActionCreator<Actions.Search.FetchAction> = (
  keyword: string,
): Actions.Search.FetchAction => ({
  type: ActionTypes.SEARCH_KEYWORD,
  keyword,
});

export const fetchStockSearchPending: ActionCreator<Actions.Search.FetchPendingAction> = (
  keyword: string,
): Actions.Search.FetchPendingAction => ({
  type: ActionTypes.SEARCH_KEYWORD_PENDING,
  keyword,
});

export const fetchStockSearchFulfilled: ActionCreator<Actions.Search.FetchFulfilledAction> = (
  keyword: string,
  payload: DataDomain.StockSearch,
): Actions.Search.FetchFulfilledAction => ({
  type: ActionTypes.SEARCH_KEYWORD_FULFILLED,
  keyword,
  payload,
});

export const fetchStockSearchRejected: ActionCreator<Actions.Search.FetchRejectedAction> = (
  keyword: string,
  error: Error,
): Actions.Search.FetchRejectedAction => ({
  type: ActionTypes.SEARCH_KEYWORD_REJECTED,
  keyword,
  error,
});

// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
//                                                                                                //
//                                  STOCK BATCH ACTIONS CREATORS                                  //
//                                                                                                //
// ////////////////////////////////////////////////////////////////////////////////////////////// //
// ////////////////////////////////////////////////////////////////////////////////////////////// //

// /**
//  * Fetch multiple stock quote data in a single batch action creator.
//  * ONLY to be called by the front end.
//  * @param stockSymbols - An array of company stock symbol in uppercase. Ex: ['AAPL', 'SHOP]
//  */
// export const fetchStockQuoteBatch: ActionCreator<Actions.Batch.FetchQuoteBatchAction> = (): Actions.Batch.FetchQuoteBatchAction => ({
//   type: ActionTypes.FETCH_STOCK_QUOTE_BATCH,
// });

// /**
//  * Fetch multiple stock chart data in a single batch action creator.
//  * ONLY to be called by the front end.
//  * @param stockSymbols - An array of company stock symbol in uppercase. Ex: ['AAPL', 'SHOP]
//  * @param range - Date range for the stock chart
//  * @param sort - sort data by date in ascending or descending order
//  */
// export const fetchStockChartBatch: ActionCreator<Actions.Batch.FetchDailyAdjustedBatchAction> = (): Actions.Batch.FetchDailyAdjustedBatchAction => ({
//   type: ActionTypes.FETCH_STOCK_DAILY_ADJUSTED_BATCH,
// });
