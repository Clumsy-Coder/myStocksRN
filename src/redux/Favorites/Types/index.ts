import ActionTypes from '@redux/Favorites/Types/ActionTypes';
import * as Actions from '@redux/Favorites/Types/Actions';
import * as Reducer from '@redux/Favorites/Types/Reducer';

export { ActionTypes };
export { Actions };
export { Reducer };
