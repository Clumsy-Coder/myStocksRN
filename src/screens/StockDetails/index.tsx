import * as React from 'react';
import { StyleSheet } from 'react-native';
import { connect } from 'react-redux';
import { Container, Content, Text, H1, View, Button, Spinner } from 'native-base';
import { Dispatch } from 'redux';

import { StockDetailsRouteProp } from '@routes/Types';
import { AppState, AppActions } from '@redux/index.reducers';
import { Actions, Selectors, DataDomain } from '@redux/Stocks/Types';
import { fetchStockChart } from '@redux/Stocks/Actions';
import { selectStockDetailsTrim, selectStockDetailsLineChart } from '@redux/Stocks/Selectors';
import { defaultQuote } from '@redux/Stocks/Reducer';
import StocksKeyStats from '@components/StocksKeyStats';
import LineChartView from '@components/LineChart';

interface SelectorProps {
  data: Selectors.SelectStockDetailsTrim;
  lineChartData: { date: string; price: number }[];
}

interface DispatchProps {
  fetchChart: (symbol: string, chartRange: DataDomain.ChartRange) => Actions.Chart.FetchAction;
}

type NavigationProps = {
  route: StockDetailsRouteProp;
};

type Props = NavigationProps & SelectorProps & DispatchProps;

type State = {
  chartRange: DataDomain.ChartRange;
};

const styles = StyleSheet.create({
  root: {
    flex: 1,
    flexDirection: 'row',
    padding: 10,
  },
  price: {
    fontWeight: 'bold',
    fontSize: 40,
    paddingTop: 20,
  },
  chartRange: {
    flex: 1,
    flexDirection: 'row',
    flexWrap: 'nowrap',
    justifyContent: 'space-between',
    paddingBottom: 10,
    flexGrow: 1,
    flexBasis: '100%',
  },
  loadingView: {
    flex: 1,
    justifyContent: 'center',
    alignContent: 'center',
    alignItems: 'center',
  },
});

export class StockDetailsScreen extends React.Component<Props, State> {
  private fetchChartInterval: NodeJS.Timeout | undefined;

  private fetchChartIntervalTimeout = 10;

  constructor(props: Props) {
    super(props);
    this.state = {
      chartRange: DataDomain.ChartRange.ONE_MONTH,
    };
  }

  componentDidMount = (): void => {
    const { route, fetchChart } = this.props;
    const { chartRange } = this.state;

    fetchChart(route.params.symbol, chartRange); // setInterval won't fetch chart. It will wait until time runs out.
    this.fetchChartInterval = setInterval(
      fetchChart.bind(this, route.params.symbol, chartRange),
      this.fetchChartIntervalTimeout * 1000,
    );
  };

  componentDidUpdate = (prevProps: Props, prevState: State): void => {
    const { chartRange } = this.state;

    if (prevState.chartRange !== chartRange) {
      const { route, fetchChart } = this.props;

      fetchChart(route.params.symbol, chartRange);

      this.fetchChartInterval = setInterval(
        fetchChart.bind(this, route.params.symbol, chartRange),
        this.fetchChartIntervalTimeout * 1000,
      );
    }
  };

  componentWillUnmount = (): void => {
    if (this.fetchChartInterval) clearInterval(this.fetchChartInterval);
  };

  onChartRangeChange = (newChartRange: DataDomain.ChartRange): void => {
    if (this.fetchChartInterval) clearInterval(this.fetchChartInterval);
    this.setState({ chartRange: newChartRange });
  };

  render(): JSX.Element {
    const { data, lineChartData } = this.props;
    const { chartRange } = this.state;

    const chartRangeList: DataDomain.ChartRange[] = [
      DataDomain.ChartRange.FIVE_DAYS,
      DataDomain.ChartRange.ONE_MONTH,
      DataDomain.ChartRange.SIX_MONTHS,
      DataDomain.ChartRange.YEAR_TO_DATE,
      DataDomain.ChartRange.ONE_YEAR,
      DataDomain.ChartRange.FIVE_YEARS,
      DataDomain.ChartRange.MAX,
    ];

    if (data.fetching && data.quote === defaultQuote && data.chart.length === 0) {
      return (
        <Container>
          <Content>
            <Spinner />
          </Content>
        </Container>
      );
    }

    return (
      <Container style={styles.root}>
        <Content>
          <View>
            <H1>{data.quote?.companyName}</H1>
          </View>
          <H1 style={styles.price}>{data.quote.latestPrice.toFixed(2)}</H1>
          <LineChartView data={lineChartData} />
          <View style={styles.chartRange}>
            {chartRangeList.map((cur) => (
              <Button
                key={`chartRange-${cur.toUpperCase()}`}
                bordered={chartRange === cur ? undefined : true}
                onPress={(): void => this.onChartRangeChange(cur)}
              >
                <Text>{cur.toUpperCase()}</Text>
              </Button>
            ))}
          </View>
          <StocksKeyStats data={data.quote} />
        </Content>
      </Container>
    );
  }
}

export const mapStateToProps = (state: AppState, ownProps: Props): SelectorProps => ({
  data: selectStockDetailsTrim(state, {
    stockSymbol: ownProps.route.params.symbol,
  }),
  lineChartData: selectStockDetailsLineChart(state, {
    stockSymbol: ownProps.route.params.symbol,
  }),
});

export const mapDispatchToProps = (dispatch: Dispatch<AppActions>): DispatchProps => ({
  fetchChart: (symbol: string, chartRange: DataDomain.ChartRange): Actions.Chart.FetchAction =>
    dispatch(fetchStockChart(symbol, chartRange)),
});

export default connect(mapStateToProps, mapDispatchToProps)(StockDetailsScreen);
