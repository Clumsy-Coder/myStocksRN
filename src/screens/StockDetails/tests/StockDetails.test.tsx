import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import { RouteProp } from '@react-navigation/native';
import { Container, Content, Spinner } from 'native-base';

import { AppState } from 'src/redux/index.reducers';
import { Selectors, DataDomain } from '@redux/Stocks/Types';
import { fetchStockChart } from '@redux/Stocks/Actions';
import { defaultQuote } from '@redux/Stocks/Reducer';
import { StockDetailsScreen, mapStateToProps, mapDispatchToProps } from '@screens/StockDetails';
import { NavigationRoutePath, StockDetailsRouteProp } from '@routes/Types';

import * as testdata from 'jest.testdata';

const selectStockDetailsTrim1: Selectors.SelectStockDetailsTrim = {
  fetching: false,
  quote: testdata.stockQuoteData1,
  chart: testdata.stockChartData1,
};

const selectStockDetailsTrim2: Selectors.SelectStockDetailsTrim = {
  fetching: false,
  quote: testdata.stockQuoteData1,
  chart: testdata.stockChartData1,
};

const lineChartData1: { date: string; price: number }[] = testdata.stockChartData1.map((cur) => ({
  date: cur.date,
  price: cur.close,
}));

const lineChartData2: { date: string; price: number }[] = testdata.stockChartData2.map((cur) => ({
  date: cur.date,
  price: cur.close,
}));

describe('<StockDetailsScreen />', () => {
  const route: StockDetailsRouteProp = {
    key: '',
    name: NavigationRoutePath.StockDetails,
    params: {
      symbol: testdata.stockSymbol1,
    },
  };
  it('Renders correctly', () => {
    const props = {
      data: selectStockDetailsTrim1,
      lineChartData: lineChartData1,
      fetchChart: jest.fn(),
      route,
    };

    const wrapper = shallow(<StockDetailsScreen {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe.skip('componentDidMount', () => {});

  describe.skip('componentWillUnmount', () => {});

  describe.skip('componentDidUpdate', () => {});

  describe('loading spinner', () => {
    it('should display spinner when loading data', () => {
      const props = {
        route,
        data: { fetching: true, quote: defaultQuote, chart: [] },
        lineChartData: [],
        fetchChart: jest.fn(),
      };

      const wrapper = shallow(<StockDetailsScreen {...props} />);

      expect(
        wrapper.equals(
          <Container>
            <Content>
              <Spinner />
            </Content>
          </Container>,
        ),
      ).toEqual(true);
    });
  });

  describe('mapStateToProps', () => {
    const rootState: AppState = {
      ...testdata.baseAppState,
      Stocks: {
        ...testdata.baseStocksState,
        symbols: {
          [testdata.stockSymbol1]: {
            chart: {
              ...testdata.baseStockChartState,
              data: testdata.stockChartData1,
            },
            quote: {
              fetching: false,
              data: testdata.stockQuoteData1,
              error: undefined,
            },
          },
        },
      },
    };

    it("property 'data' works properly", () => {
      const stateToProps = mapStateToProps(rootState, {
        data: {
          chart: [],
          fetching: false,
          quote: testdata.stockQuoteData1,
        },
        lineChartData: [],
        route: {
          key: '',
          name: NavigationRoutePath.StockDetails,
          params: {
            symbol: testdata.stockSymbol1,
          },
        },
        fetchChart: jest.fn(),
      });
      expect(stateToProps).toHaveProperty('data');
      expect(stateToProps.data).toEqual({
        fetching: false,
        quote: testdata.stockQuoteData1,
        chart: testdata.stockChartData1,
      });
    });

    it("property 'lineChartData' works properly", () => {
      const stateToProps = mapStateToProps(rootState, {
        data: {
          chart: [],
          fetching: false,
          quote: testdata.stockQuoteData1,
        },
        lineChartData: [],
        route: {
          key: '',
          name: NavigationRoutePath.StockDetails,
          params: {
            symbol: testdata.stockSymbol1,
          },
        },
        fetchChart: jest.fn(),
      });
      expect(stateToProps).toHaveProperty('lineChartData');
      expect(stateToProps.lineChartData).toEqual(
        testdata.stockChartData1.map((cur) => ({ date: cur.date, price: cur.close })),
      );
    });
  });

  describe('mapDispatchToProps', () => {
    it('fetchChart works properly', () => {
      const dispatch = jest.fn();

      const dispatchToProps = mapDispatchToProps(dispatch);
      dispatchToProps.fetchChart(testdata.stockSymbol1, DataDomain.ChartRange.ONE_MONTH);

      expect(dispatch.mock.calls[0].length).toEqual(1);
      expect(dispatch.mock.calls[0][0]).toEqual(
        fetchStockChart(testdata.stockSymbol1, DataDomain.ChartRange.ONE_MONTH),
      );
    });
  });
});
