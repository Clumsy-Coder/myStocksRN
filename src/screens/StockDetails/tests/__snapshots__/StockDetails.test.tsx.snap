// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StockDetailsScreen /> Renders correctly 1`] = `
<Styled(Container)
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "padding": 10,
    }
  }
>
  <Styled(Content)>
    <Styled(ViewNB)>
      <Styled(H1)>
        International Business Machines Corp.
      </Styled(H1)>
    </Styled(ViewNB)>
    <Styled(H1)
      style={
        Object {
          "fontSize": 40,
          "fontWeight": "bold",
          "paddingTop": 20,
        }
      }
    >
      122.47
    </Styled(H1)>
    <LineChartView
      data={
        Array [
          Object {
            "date": "2020-06-12",
            "price": 121.91,
          },
          Object {
            "date": "2020-06-15",
            "price": 121.65,
          },
        ]
      }
    />
    <Styled(ViewNB)
      style={
        Object {
          "flex": 1,
          "flexBasis": "100%",
          "flexDirection": "row",
          "flexGrow": 1,
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
          "paddingBottom": 10,
        }
      }
    >
      <Styled(Button)
        bordered={true}
        key="chartRange-5D"
        onPress={[Function]}
      >
        <Styled(Text)>
          5D
        </Styled(Text)>
      </Styled(Button)>
      <Styled(Button)
        key="chartRange-1M"
        onPress={[Function]}
      >
        <Styled(Text)>
          1M
        </Styled(Text)>
      </Styled(Button)>
      <Styled(Button)
        bordered={true}
        key="chartRange-6M"
        onPress={[Function]}
      >
        <Styled(Text)>
          6M
        </Styled(Text)>
      </Styled(Button)>
      <Styled(Button)
        bordered={true}
        key="chartRange-YTD"
        onPress={[Function]}
      >
        <Styled(Text)>
          YTD
        </Styled(Text)>
      </Styled(Button)>
      <Styled(Button)
        bordered={true}
        key="chartRange-1Y"
        onPress={[Function]}
      >
        <Styled(Text)>
          1Y
        </Styled(Text)>
      </Styled(Button)>
      <Styled(Button)
        bordered={true}
        key="chartRange-5Y"
        onPress={[Function]}
      >
        <Styled(Text)>
          5Y
        </Styled(Text)>
      </Styled(Button)>
      <Styled(Button)
        bordered={true}
        key="chartRange-MAX"
        onPress={[Function]}
      >
        <Styled(Text)>
          MAX
        </Styled(Text)>
      </Styled(Button)>
    </Styled(ViewNB)>
    <StocksKeyStats
      data={
        Object {
          "avgTotalVolume": 4866120,
          "change": -1.69,
          "changePercent": -1.361,
          "close": 130,
          "companyName": "International Business Machines Corp.",
          "extendedChange": 2,
          "extendedChangePercent": 0.1,
          "extendedPrice": 120,
          "high": 140,
          "latestPrice": 122.47,
          "latestTime": "June 19, 2020",
          "latestUpdate": 1592596802601,
          "latestVolume": 2865362,
          "low": 100,
          "marketCap": 108740133240,
          "open": 110,
          "peRatio": 12.05,
          "previousClose": 124.16,
          "previousVolume": 2865362,
          "primaryExchange": "New York Stock Exchange",
          "symbol": "IBM",
          "volume": 2865320,
          "week52High": 158.75,
          "week52Low": 90.56,
          "ytdChange": -9.6759,
        }
      }
    />
  </Styled(Content)>
</Styled(Container)>
`;
